### PR TITLE
Metrics Service

### PR DESCRIPTION
**Pull request by flicker-harmony**

### CODE REVIEW

**Comment by harmony-ek** [comment]
Jenkins speaking; can one of the admins verify this patch?

**Comment by LeoHChen** [comment]
Jenkins, test this please.

**Comment by flicker-harmony** [comment]
Jenkins, test this please.

**Comment by flicker-harmony** [comment]
Jenkins, test this please.

**Comment by xChickens** [comment]
ok to test

File: node/node.go [hunk 1]
```diff
@@ -236,15 +239,15 @@ func (node *Node) addPendingTransactions(newTxs types.Transactions) {
 		node.pendingTransactions = append(node.pendingTransactions, newTxs...)
 		node.reducePendingTransactions()
 		node.pendingTxMutex.Unlock()
-		utils.Logger().Info().Int("num", len(newTxs)).Int("totalPending", len(node.pendingTransactions)).Msg("Got more transactions")
+		utils.GetLogInstance().Info("Got more transactions", "num", len(newTxs), "totalPending", len(node.pendingTransactions))
```
Comment (LeoHChen) [inline node/node.go]: please change back to use the zerolog.  those code has to be reverted.

File: node/node.go [hunk 2]
```diff
@@ -236,15 +239,15 @@ func (node *Node) addPendingTransactions(newTxs types.Transactions) {
...
 	}
 }
 
 // AddPendingTransaction adds one new transaction to the pending transaction list.
 func (node *Node) AddPendingTransaction(newTx *types.Transaction) {
 	if node.NodeConfig.GetNetworkType() != nodeconfig.Mainnet {
 		node.addPendingTransactions(types.Transactions{newTx})
-		utils.Logger().Error().Int("totalPending", len(node.pendingTransactions)).Msg("Got ONE more transaction")
+		utils.GetLogInstance().Debug("Got ONE more transaction", "totalPending", len(node.pendingTransactions))
```
Comment (LeoHChen) [inline node/node.go]: same thing here.

File: node/node.go [hunk 3]
```diff
@@ -259,11 +262,7 @@ func (node *Node) getTransactionsForNewBlock(maxNumTxs int, coinbase common.Addr
 
 	node.pendingTransactions = unselected
 	node.reducePendingTransactions()
-	utils.Logger().Error().
-		Int("remainPending", len(node.pendingTransactions)).
-		Int("selected", len(selected)).
-		Int("invalidDiscarded", len(invalid)).
-		Msg("Selecting Transactions")
+	utils.GetLogInstance().Debug("Selecting Transactions", "remainPending", len(node.pendingTransactions), "selected", len(selected), "invalidDiscarded", len(invalid))
```
Comment (LeoHChen) [inline node/node.go]: use zerolog

File: node/node.go [hunk 4]
```diff
@@ -364,14 +363,13 @@ func New(host p2p.Host, consensusObj *consensus.Consensus, chainDBFactory shardc
 			var err error
 			node.TestBankKeys, err = CreateTestBankKeys(TestAccountNumber)
 			if err != nil {
-				utils.Logger().Error().Err(err).Msg("Error while creating test keys")
+				utils.GetLogInstance().Crit("Error while creating test keys",
```
Comment (LeoHChen) [inline node/node.go]: use zero log.

File: node/node.go [hunk 5]
```diff
@@ -364,14 +363,13 @@ func New(host p2p.Host, consensusObj *consensus.Consensus, chainDBFactory shardc
...
+					"error", err)
 			}
 		}
 	}
 
-	utils.Logger().Info().
-		Interface("genesis block header", node.Blockchain().GetBlockByNumber(0).Header()).
-		Msg("Genesis block hash")
+	utils.GetLogInstance().Info("Genesis block hash", "genesis block header", node.Blockchain().GetBlockByNumber(0).Header())
```
Comment (LeoHChen) [inline node/node.go]: zerolog

File: node/node.go [hunk 6]
```diff
@@ -486,17 +489,17 @@ func (node *Node) initNodeConfiguration() (service.NodeConfig, chan p2p.Peer) {
 	var err error
 	node.shardGroupReceiver, err = node.host.GroupReceiver(node.NodeConfig.GetShardGroupID())
 	if err != nil {
-		utils.Logger().Error().Err(err).Msg("Failed to create shard receiver")
+		utils.GetLogInstance().Error("Failed to create shard receiver", "msg", err)
```
Comment (LeoHChen) [inline node/node.go]: zerolog

File: node/node.go [hunk 7]
```diff
@@ -486,17 +489,17 @@ func (node *Node) initNodeConfiguration() (service.NodeConfig, chan p2p.Peer) {
...
 	node.globalGroupReceiver, err = node.host.GroupReceiver(p2p.GroupIDBeaconClient)
 	if err != nil {
-		utils.Logger().Error().Err(err).Msg("Failed to create global receiver")
+		utils.GetLogInstance().Error("Failed to create global receiver", "msg", err)
 	}
 
 	node.clientReceiver, err = node.host.GroupReceiver(node.NodeConfig.GetClientGroupID())
 	if err != nil {
-		utils.Logger().Error().Err(err).Msg("Failed to create client receiver")
+		utils.GetLogInstance().Error("Failed to create client receiver", "msg", err)
```
Comment (LeoHChen) [inline node/node.go]: zerolog pls

File: node/node_metrics.go [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+)
+
+// UpdateBlockHeightForMetrics updates block height for metrics service.
+func (node *Node) UpdateBlockHeightForMetrics(prevBlockHeight uint64) uint64 {
+	curBlock := node.Blockchain().CurrentBlock()
+	curBlockHeight := curBlock.NumberU64()
+	if curBlockHeight == prevBlockHeight {
+		return prevBlockHeight
+	}
+	log.Info().Msgf("Updating metrics block height %d", curBlockHeight)
```
Comment (LeoHChen) [inline node/node_metrics.go]: use utils.Logger() please.

File: node/service_setup.go [hunk 9]
```diff
@@ -96,7 +152,7 @@ func (node *Node) ServiceManagerSetup() {
 // RunServices runs registered services.
 func (node *Node) RunServices() {
 	if node.serviceManager == nil {
-		utils.Logger().Info().Msg("Service manager is not set up yet.")
+		utils.GetLogInstance().Info("Service manager is not set up yet.")
```
Comment (LeoHChen) [inline node/service_setup.go]: zerolog

File: node/service_setup.go [hunk 10]
```diff
@@ -28,13 +29,64 @@ func (node *Node) setupForValidator() {
 	node.serviceManager.RegisterService(service.BlockProposal, blockproposal.New(node.Consensus.ReadySignal, node.WaitForConsensusReadyv2))
 	// Register client support service.
 	node.serviceManager.RegisterService(service.ClientSupport, clientsupport.New(node.Blockchain().State, node.CallFaucetContract, node.getDeployedStakingContract, node.SelfPeer.IP, node.SelfPeer.Port))
+	// Register new metrics service
+	node.serviceManager.RegisterService(service.Metrics, metrics.New(&node.SelfPeer, node.NodeConfig.ConsensusPubKey.SerializeToHexStr(), node.NodeConfig.GetPushgatewayIP(), node.NodeConfig.GetPushgatewayPort(), node.Consensus.GetNodeIDs))
+}
+
+func (node *Node) setupForShardValidator() {
```
Comment (LeoHChen) [inline node/service_setup.go]: those code are old code should be removed. you don't have the latest code.

**Comment by flicker-harmony** [comment]
Jenkins, test this please.

**Pull request merged**